zerver/lib/bugdown/__init__.py

```diff
@@ -1001,6 +1001,36 @@ def run(self, root: Element) -> None:
                 else:
                     add_embed(root, url, extracted_data)
 
+class CleanUpProcessor(markdown.treeprocessors.Treeprocessor):
+    """Applies final logic related to user mentions after all the other processors"""
+
+    def handle_quoted_user_mentions(self, root: Element) -> Element:
+        if not self.markdown.zulip_message:
+            return root
+
+        # Just keep the HTML for styling purposes for mentions inside blockquotes.
+        for quote in walk_tree(root, lambda e: e if e.tag == "blockquote" else None):
+            for mention_tag in walk_tree(quote, lambda e: e if e.get('class') == "user-mention" else None):
+                current_id = mention_tag.get('data-user-id')
+                if current_id:
+                    mention_tag.set('data-silent-user-id', current_id)
+                mention_tag.set('data-user-id', '')
+
+        # Recalculate mentions_user_ids field with only the ids that are mentioned outside quotes.
+        ids_mentioned_outside_quotes = set()
+        for quote in walk_tree(root, lambda e: e if not e.tag == "blockquote" else None):
+            for mention_tag in walk_tree(quote, lambda e: e if e.get('class') == "user-mention" else None):
+                user_id = mention_tag.get('data-user-id')
+                if user_id and not user_id == '*':
+                    ids_mentioned_outside_quotes.add(int(user_id))
+        self.markdown.zulip_message.mentions_user_ids = ids_mentioned_outside_quotes
+
+        return root
```

Comment:
Did you try just walking parents from the node where a `mention` item is being written out to see if it's inside a blockquote?  I think the order-of-operations might be right for this.

Reply:
If note, this is a reasonable approach, though I'm suspicious of the nested `walk_tree` and we'll want to replace that with a smarter walk that just keeps track of whether the current node is a descendant of a blockquote.

I think it might also be a good refactoring of part of this to just construct `self.markdown.zulip_message.mentions_user_ids` here, rather than in `UserMentionPattern`, to avoid needing to do that work twice.  Less confusing.

Reply:
I tried but it seems that there isn't any non-roundabout way to access any elements of the tree from an InlinePattern. I'll read the code for the inline TreeProcessor again, but there doesn't seem to be any way outside of subclassing the inline treeprocessor.